clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2882,12 +2889,16 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
     // (of either the lambda or the function object).
     CXXRecordDecl *KernelObj =
         GetSYCLKernelObjectType(KernelCallerFunc)->getAsCXXRecordDecl();
-    CXXMethodDecl *WGLambdaFn = nullptr;
+
+    KernelCallOperatorVisitor KernelCallOperator(KernelCallerFunc, KernelObj);
+    CXXMethodDecl *WGCallOperator = nullptr;
+
     if (KernelObj->isLambda())
-      WGLambdaFn = KernelObj->getLambdaCallOperator();
-    else
-      WGLambdaFn = getOperatorParens(KernelObj);
-    assert(WGLambdaFn && "non callable object is passed as kernel obj");
+      WGCallOperator = KernelObj->getLambdaCallOperator();
```

Comment:
So, this code is inside the body creator, meaning the traverse has happened earlier before the visitor invoked. Can we just pass call operator as constructor parameter to the visitor to not traverse twice?

Reply:
I have updated the code to pass `CallOperator` to `SyclKernelBodyCreator`, `copySYCLKernelAttrs `and `isESIMDKernelType`